src/code/audio/audio_seqplayer.c

```diff
@@ -1,73 +1,2314 @@
+    layer->unk_08 = 0x80;
+    layer->stereo.asByte = 0;
+    layer->portamento.mode = PORTAMENTO_MODE_OFF;
+    layer->scriptState.depth = 0;
+    layer->pan = 0x40;
+    layer->transposition = 0;
+    layer->delay = 0;
+    layer->gateDelay = 0;
+    layer->delay2 = 0;
+    layer->note = NULL;
+    layer->instrument = NULL;
+    layer->instOrWave = 0xFF;
+    layer->unk_0A.asByte = 0xFFFF;
+    layer->vibrato.vibratoRateTarget = 0x800;
+    layer->vibrato.vibratoRateStart = 0x800;
+    layer->vibrato.vibratoExtentTarget = 0;
+    layer->vibrato.vibratoExtentStart = 0;
+    layer->vibrato.vibratoRateChangeDelay = 0;
+    layer->vibrato.vibratoExtentChangeDelay = 0;
+    layer->vibrato.vibratoDelay = 0;
+    layer->freqScale = 1.0f;
+    layer->bend = 1.0f;
+    layer->velocitySquare2 = 0.0f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_seqplayer/func_8019A0BC.s")
+    return 0;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_seqplayer/func_8019AA3C.s")
+void AudioSeq_SeqLayerDisable(SequenceLayer* layer) {
+    if (layer != NULL) {
+        if ((layer->channel != &gAudioContext.sequenceChannelNone) && (layer->channel->seqPlayer->finished == true)) {
+            AudioPlayback_SeqLayerNoteRelease(layer);
+        } else {
+            AudioPlayback_SeqLayerNoteDecay(layer);
+        }
+        layer->enabled = false;
+        layer->finished = true;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_seqplayer/func_8019AAF0.s")
+void AudioSeq_SeqLayerFree(SequenceChannel* channel, s32 layerIndex) {
+    SequenceLayer* layer = channel->layers[layerIndex];
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_seqplayer/AudioSeq_ResetSequencePlayer.s")
+    if (layer != NULL) {
+        AudioSeq_AudioListPushBack(&gAudioContext.layerFreeList, &layer->listItem);
+        AudioSeq_SeqLayerDisable(layer);
+        channel->layers[layerIndex] = NULL;
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_seqplayer/AudioSeq_InitSequencePlayerChannels.s")
+void AudioSeq_SequenceChannelDisable(SequenceChannel* channel) {
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_seqplayer/func_8019ACEC.s")
+    channel->finished = true;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/audio_seqplayer/AudioSeq_InitSequencePlayers.s")
+    for (i = 0; i < 4; i++) {
+        AudioSeq_SeqLayerFree(channel, i);
+    }
+
+    AudioPlayback_NotePoolClear(&channel->notePool);
+    channel->enabled = false;
+}
+
+void AudioSeq_SequencePlayerSetupChannels(SequencePlayer* seqPlayer, u16 channelBits) {
+    SequenceChannel* channel;
+    s32 i;
+
+    for (i = 0; i < ARRAY_COUNT(seqPlayer->channels); i++) {
+        if (channelBits & 1) {
+            channel = seqPlayer->channels[i];
+            channel->fontId = seqPlayer->defaultFont;
+            channel->muteBehavior = seqPlayer->muteBehavior;
+            channel->noteAllocPolicy = seqPlayer->noteAllocPolicy;
+        }
+        channelBits = channelBits >> 1;
```

Comment:
Are channelBits set for channels we are going to enable or left disabled/ignored? I don't see why it's setting a default font, initializing the channel should handle that during channel init?

Reply:
I'm not exactly following what this comment refers to. Something to note is that I've never looked into `defaultFont` so it might not be the best name, so take that with a grain of salt. I'll add it to my audio todo list